driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorPublisher.java

```diff
@@ -0,0 +1,169 @@
+    private final MongoOperationPublisher<T> mongoOperationPublisher;
+
+    private Integer batchSize;
+
+    BatchCursorPublisher(@Nullable final ClientSession clientSession, final MongoOperationPublisher<T> mongoOperationPublisher) {
+        this.clientSession = clientSession;
+        this.mongoOperationPublisher = notNull("mongoOperationPublisher", mongoOperationPublisher);
+    }
+
+    abstract AsyncReadOperation<AsyncBatchCursor<T>> asAsyncReadOperation();
+
+    AsyncReadOperation<AsyncBatchCursor<T>> asAsyncFirstReadOperation() {
+        return asAsyncReadOperation();
+    }
+
+    @Nullable
+    ClientSession getClientSession() {
+        return clientSession;
+    }
+
+    MongoOperationPublisher<T> getMongoOperationPublisher() {
+        return mongoOperationPublisher;
+    }
+
+    Operations<T> getOperations() {
+        return mongoOperationPublisher.getOperations();
+    }
+
+    MongoNamespace getNamespace() {
+        return mongoOperationPublisher.getNamespace();
+    }
+
+    ReadPreference getReadPreference() {
+        return mongoOperationPublisher.getReadPreference();
+    }
+
+    CodecRegistry getCodecRegistry() {
+        return mongoOperationPublisher.getCodecRegistry();
+    }
+
+    boolean getRetryReads() {
+        return mongoOperationPublisher.getRetryReads();
+    }
+
+    Class<T> getDocumentClass() {
+        return mongoOperationPublisher.getDocumentClass();
+    }
+
+
+    @Nullable
+    public Integer getBatchSize() {
+        return batchSize;
+    }
+
+    public Publisher<T> batchSize(final int batchSize) {
+        this.batchSize = batchSize;
+        return this;
+    }
+
+    public Publisher<T> first() {
+        return batchCursor(this::asAsyncFirstReadOperation)
+                .flatMap(batchCursor -> Mono.create(sink -> {
+                    batchCursor.setBatchSize(1);
+                    Mono.from(batchCursor.next())
+                            .doOnTerminate(batchCursor::close)
+                            .doOnError(sink::error)
+                            .doOnSuccess(results -> {
+                                if (results == null || results.isEmpty()) {
+                                    sink.success();
+                                } else {
+                                    sink.success(results.get(0));
+                                }
+                            })
+                            .subscribe();
+                }));
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super T> subscriber) {
+        batchCursor()
+                .flatMapMany(batchCursor -> {
+                    AtomicBoolean inProgress = new AtomicBoolean(false);
+                    if (batchSize != null) {
+                        batchCursor.setBatchSize(batchSize);
+                    }
+                    return Flux.create((FluxSink<T> sink) -> {
+                        sink.onRequest(value -> recurseCursor(sink, batchCursor, inProgress));
```

Comment:
I'm confused by the three different places `recurseCursor` is invoked.  Once in this lambda, once just below, and then once in the lambda in `recurseCursor`  Can you explain what's going on here?

Reply:
There is an extra call here to recurseCursor that isn't needed as all demand will be requested via the sink.

If a user signals for more data from the publisher - we need to return the results from the cursor and potentially fetch more.  `sink.onRequest` ensures that demand from the user is satisfied.

However a user may signal demand any number of times, so we can't just rely on using recurseCursor onRequest - so we add a lock (inProgress).  Once we have got results we process them,  then we recurse to check if there is any outstanding demand from the user.